sign/dilithium/mode1/dilithium.go

```diff
@@ -0,0 +1,128 @@
+// mode1 implements the CRYSTALS-Dilithium signature scheme Dilithium1
+// as submitted to round2 of the NIST PQC competition and described in
+//
+// https://pq-crystals.org/dilithium/data/dilithium-specification-round2.pdf
+package mode1
+
+import (
+	"errors"
+	"io"
+
+	common "github.com/cloudflare/circl/sign/dilithium/internal"
+	"github.com/cloudflare/circl/sign/dilithium/mode1/internal"
+)
+
+const (
+	// Size of seed for NewKeyFromSeed
+	SeedSize = common.SeedSize
+
+	// Size of a packed PublicKey
+	PublicKeySize = internal.PublicKeySize
+
+	// Size of a packed PrivateKey
+	PrivateKeySize = internal.PrivateKeySize
+
+	// Size of a signature
+	SignatureSize = internal.SignatureSize
+)
+
+// PublicKey is the type of Dilithium1 public key
+type PublicKey internal.PublicKey
+
+// PrivateKey is the type of Dilithium1 private key
+type PrivateKey internal.PrivateKey
+
+// GenerateKey generates a public/private key pair using entropy from rand.
+// If rand is nil, crypto/rand.Reader will be used.
+func GenerateKey(rand io.Reader) (*PublicKey, *PrivateKey, error) {
+	pk, sk, err := internal.GenerateKey(rand)
+	return (*PublicKey)(pk), (*PrivateKey)(sk), err
+}
+
+// NewKeyFromSeed derives a public/private key pair using the given seed.
+func NewKeyFromSeed(seed *[SeedSize]byte) (*PublicKey, *PrivateKey) {
+	pk, sk := internal.NewKeyFromSeed(seed)
+	return (*PublicKey)(pk), (*PrivateKey)(sk)
+}
+
+// SignTo signs the given message and writes the signature into signature.
+// It will panic if signature is not of length at least SignatureSize.
+func SignTo(sk *PrivateKey, msg []byte, signature []byte) {
```

Comment:
Both Sign and SignTo have the same functionality. I understand the advantages of one over the other, however it looks to me as duplication of code. That's why I recommend to use one of them.

> So two is fine if one of them is `crypto.Signer` or would only implement `crypto.Signer`.

Having something that implements crypto.Signer, it would be an additional feature to have.

Reply:
I implemented the `crypto.Signer` interface.

Reply:
I removed `SignTo()` from `Mode` (as allocations can't really be avoided with interfaces) and removed `Sign()` in the mode subpackages (as one would use these when trying to get the maximum performance anyway).